plugin-server/src/main/ingestion-queues/anonymous-event-buffer-consumer.ts

```diff
@@ -0,0 +1,97 @@
+import { PluginEvent } from '@posthog/plugin-scaffold'
+import { Kafka } from 'kafkajs'
+import { JobQueueManager } from 'main/job-queues/job-queue-manager'
+import { KafkaProducerWrapper } from 'utils/db/kafka-producer-wrapper'
+
+import { KAFKA_BUFFER, KAFKA_EVENTS_DEAD_LETTER_QUEUE } from '../../config/kafka-topics'
+import { JobName } from '../../types'
+import { status } from '../../utils/status'
+
+export const startAnonymousEventBufferConsumer = async ({
+    kafka,
+    producer,
+    jobQueueManager,
+}: {
+    kafka: Kafka
+    producer: KafkaProducerWrapper
+    jobQueueManager: JobQueueManager
+}) => {
+    /*
+        Consumes from the anonymous event topic, and enqueues the events for
+        processing at a later date as per the message header `processEventAt`.
+
+        We do this delayed processing to allow for the anonymous users that
+        these events are assosiated to be merged or identified as other
+        "identified" users, at which point to can denormalize data to improve
+        query performance.
+
+        On failure to enqueue we will fail the batch. only on successfully
+        enqueuing an entire batch to Graphile Worker (a PostgreSQL backed async
+        worker library).
+
+        TODO:
+
+        1. distinguish between operational errors and programming errors
+        2. ensure offset only updated if message pushed to DQL
+    */
+
+    const consumer = kafka.consumer({ groupId: 'ingester' })
+
+    status.info('🔁', 'Starting anonymous event buffer consumer')
+
+    await consumer.connect()
+    await consumer.subscribe({ topic: KAFKA_BUFFER })
+    await consumer.run({
+        eachBatch: async ({ batch, resolveOffset, heartbeat }) => {
+            status.info('🔁', 'Processing batch', { size: batch.messages.length })
+            for (const message of batch.messages) {
+                if (!message.value || !message.headers?.processEventAt || !message.headers?.eventId) {
+                    status.warn(`Invalid message for partition ${batch.partition} offset ${message.offset}.`)
+                    producer.queueMessage({ topic: KAFKA_EVENTS_DEAD_LETTER_QUEUE, messages: [message] })
+                    resolveOffset(message.offset)
+                    continue
+                }
+
+                let eventPayload: PluginEvent
+
+                try {
+                    eventPayload = JSON.parse(message.value.toString())
+                } catch (error) {
+                    status.warn(`Invalid message for partition ${batch.partition} offset ${message.offset}.`)
```

Comment:
should we change the warn message from above? This is a more specific case of json unmarshaling error

Reply:
Have included the error